### PR TITLE
[WIP] feat: descending iterations

### PR DESCRIPTION
**Pull request by rzvxa**
When I find some free time; I'll clean this up and add iterator support then mark it for review. I should also add more tests to make sure everything are behaving as expected.

related to #2 and #3 

Please feel free to give it an early review in case I'm doing it wrong, I've opt-in for monomorphic code and reducing the indirections by copying the whole iteration logic(including the contexts), It might not be as desirable as having a runtime check(in terms of tidy code) but it'd be nice to avoid extra branches in such a hot path.

### CODE REVIEW

File: tree_traversal.go [hunk 1]
```diff
@@ -53,6 +53,36 @@ func traversalContext4_16_256(childZeroIdx int) *traverse4_16_256 {
...
+// traverseDesc4_16_256 is a context for traversing nodes with 4, 16, or 256 children in descending order.
+type traverseDesc4_16_256 struct {
+	numChildren     int
+	childZeroActive bool
+	childCurIdx     int
+}
+
+// nextChildIdx returns the index of the next child node to traverse.
+func (tc *traverseDesc4_16_256) nextChildIdx() (int, bool) {
+	if tc.childZeroActive {
```
Comment (rzvxa) [inline tree_traversal.go:65]: TBH, I'm not sure about this part, I've given the spec a read but I've discovered ART fairly recently so I might be missing something here.

**Comment by plar** [comment]
Hey Rez! What are you aiming to achieve with this implementation? Is your goal to enable iteration over the same tree in both forward and reverse directions? Or would it be sufficient for you to create a tree that inherently supports reverse-order traversal?

**Comment by plar** [comment]
@rzvxa Could you review this [PR](https://github.com/plar/go-adaptive-radix-tree/pull/33)? Does it address your requirements?

**Comment by plar** [comment]
I am closing this PR, since this functionality was implemented in #33.

**Comment by rzvxa** [comment]
First off sorry for not being available, I was busy at work and didn't really have the time to check my notifications.
The implementation introduced in #33 is excellent, Awesome job🚀 Sorry for not being able to contribute to that PR.

Cheers❤️

**Pull request closed without merging**